### PR TITLE
fix doctests and add missing docstrings

### PR DESCRIPTION
**Pull request by kleinschmidt**
This fixes up the doctests a bit, which had gotten out of sync with the actual
package and with how GLM prints coef tables.  The most substantive changes are
to add some docstrings that were missing (`response` and `modelmatrix`) and to
update the docstring and examples for `hypothesis_matrix` to reflect the fact
that this now directly returns the pseudo-inverse (where the interpretation of
rows and columns is switched from the contrasts matrix).

### CODE REVIEW

**Comment by kleinschmidt** [comment]
I'll rebase and merge this if there are no objections.

File: docs/src/formula.md
Comment (palday) [inline docs/src/formula.md:380]: Not just the rounding but the actual values changed here. Is that correct?

File: docs/src/internals.md
Comment (palday) [inline docs/src/internals.md:528]: likewise here

File: docs/src/internals.md
Comment (palday) [inline docs/src/internals.md:638]: and here

File: docs/src/internals.md
Comment (palday) [inline docs/src/internals.md:742]: and here, although less extreme

File: docs/src/internals.md
Comment (palday) [inline docs/src/internals.md:756]: another one

File: src/contrasts.jl
Comment (palday) [inline src/contrasts.jl:670]: Yep, super important to make this clear. :+1:

**Review by palday — COMMENTED** [review]
Not just the rounding but the values of a lot of the coefficient tables changed. Was this intentional?

**Comment by kleinschmidt** [comment]
I think the different coefficients are due to the docs being built on 1.5 instead of 1.3 previously and relying on RNG...

**Comment by palday** [comment]
Ah yeah, the RNG changes are a pain.

**Comment by palday** [comment]
Can you change the RNG to use `StableRNGs`? Then we won't have these changes in the future. Either way, go ahead and rebase.

**Comment by kleinschmidt** [comment]
I thought about that but it would require tweaking the set-up too and for some reason didn't seem worth it, although now I can't quite remember why.  I'll have another look...

**Comment by kleinschmidt** [comment]
Ah I remember what it was: I'd have to also update the doctests to also take the `rng` argument.  That's not a huge deal I guess especially since it only touches the data generation part.

**Comment by palday** [comment]
I didn't comment on it, but I was curious why you set the seed in the setup because then everything else depends on it always being executed in that order and with no intervening calls to the RNG.

**Comment by kleinschmidt** [comment]
The setup block runs for every doc test (except when there's a labeled block in which case the order does matter)

**Comment by palday** [comment]
Ah, good to know!

**Review by palday — APPROVED** [review]

**Comment by palday** [comment]
Still got a merge conflict but otherwise LGTM

**Comment by kleinschmidt** [comment]
I still gotta fix the doctests...just wanted to run CI and see what happened here :)

**Comment by kleinschmidt** [comment]
Okay, a few doctests do fail, but they're differences in the path between the builder and my local path, and some strange differences in very small numbers that are only printed to illustrate that they're basically zero and are therefore distracting...

**Comment by palday** [comment]
@ararslan or @nalimilan should comment on how important they find such things.

**Comment by kleinschmidt** [comment]
yeah it would be a matter of adding a regex to ignore those failures...I don't know why the smol numbers are different given that it's deterministic input and just a matrix pseudoinverse.

**Comment by palday** [comment]
Let me tell the horrors of platform-dependence numerical linear algebra and the holy mess that incurred in rank deficiency detection in MixedModels ..... 

::wimpers::

**Comment by ararslan** [comment]
> @ararslan or @nalimilan should comment on how important they find such things.

This is literally the only thing that matters and this is a hill I'm willing to die on.

(Actually I don't care at all)

**Comment by palday** [comment]
I don't really care per se, but I do know that if you get used to ignoring an 'alarm'/warning, you wind up ignoring it when you shouldn't. The bot form for the boy who cried wolf.

**Comment by ararslan** [comment]
I would suggest that any doctests that are susceptible to numerical funny business either be converted from `jldoctest` to `julia-repl` or weaseled such that some transformation is applied as the last step that gets displayed.

**Comment by kleinschmidt** [comment]
Here's the contents of that doctest:

````
```jldoctest
julia> cmat = StatsModels.contrasts_matrix(DummyCoding(), 1, 4)
4×3 Array{Float64,2}:
 0.0  0.0  0.0
 1.0  0.0  0.0
 0.0  1.0  0.0
 0.0  0.0  1.0

julia> StatsModels.hypothesis_matrix(cmat)
4×4 Array{Int64,2}:
  1  0  0  0
 -1  1  0  0
 -1  0  1  0
 -1  0  0  1

julia> StatsModels.hypothesis_matrix(cmat, intercept=false) # wrong without intercept!!
3×4 Array{Int64,2}:
 0  1  0  0
 0  0  1  0
 0  0  0  1

julia> StatsModels.hypothesis_matrix(cmat, tolerance=0) # ugly
4×4 Array{Float64,2}:
  1.0  -2.23753e-16   6.91749e-18  -1.31485e-16
 -1.0   1.0          -2.42066e-16   9.93754e-17
 -1.0   4.94472e-17   1.0           9.93754e-17
 -1.0   1.04958e-16  -1.31044e-16   1.0        

julia> StatsModels.hypothesis_matrix(StatsModels.ContrastsMatrix(DummyCoding(), ["a", "b", "c", "d"]))
4×4 Array{Int64,2}:
  1  0  0  0
 -1  1  0  0
 -1  0  1  0
 -1  0  0  1

```
````

**Comment by kleinschmidt** [comment]
cf. `# ugly` on that line...

**Comment by ararslan** [comment]
Could have a doctest dependency on ZChop and just call `zchop` to force the tiny bits to zero, or just like `round.(Int, _)`...

**Comment by kleinschmidt** [comment]
That's basically what the `tolerance` kwarg does (clips things to `Int` or `Rational`).  The point of the example is to show what happens when you set it to zero (just passes the floats through as is).

**Comment by palday** [comment]
@kleinschmidt Maybe show that it's not exactly (elementwise) equal to 0, ±1, but that it is less than an epsilon away from that.

**Comment by nalimilan** [comment]
Sounds good. Though doctests are only run on a given Julia version so we could also hardcode whatever result we obtain there, right?

**Comment by kleinschmidt** [comment]
I thought I _was_ running it on the same version locally (1.5.3) but maybe not, or maybe it depends somehow on the architecture it's running on, who knows.  I think replacing the `jldoctest` with `juliarepl` block is a good solution.

**Comment by kleinschmidt** [comment]
Oh good the printing of type parameters (now there are spaces printed after each comma) and matrix/vector types (`Matrix{T}` instead of `Array{T,2}`) is now different on nightly so a million doctests are broken again.  I think at this point I should just give up supporting both 1.6 and 1.5 in the doctests and make another fix once 1.6 is actually released.

**Pull request merged**
